### PR TITLE
LPS-201494 Add keyboard DnD in Segments Editor

### PR DESCRIPTION
**Pull request by sandrodw3**
## Dear Reviewer

I'm sorry ❤️ (but I'm very proud of this PR 😌)

### CODE REVIEW

**Comment by sandrodw3** [comment]
ci:test:sf

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/searchAndUpdateCriteria.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+export default function searchAndUpdateCriteria(
+	criteria: Criteria | CriteriaItem,
+	startGroupId: Criteria['groupId'],
+	startIndex: number,
+	destGroupId: Criteria['groupId'],
+	destIndex: number,
+	addedItem: Criteria | CriteriaItem,
+	replace: boolean
+): Criteria {
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/searchAndUpdateCriteria.ts:17]: Same here, this function receives a bunch of parameters but I am not sure about what "startGroupId" or "destGroupId" do. Does it move an item from one place to another?

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/getGroup.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import {Criteria} from '../../types/Criteria';
+
+export default function getGroup(
+	groupId: Criteria['groupId'],
+	criteria: Criteria
+): Criteria | undefined {
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/getGroup.ts:11]: This issue was already there, but it is a bit confusing that sometimes we use both "Group" and "Criteria". This function for example returns a "Group", but the type says that this is a "Criteria".

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/getItem.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import getGroup from './getGroup';
+
+export default function getItem(
+	criteria: Criteria,
+	groupId: Criteria['groupId'],
+	index: number
+) {
+	const group = getGroup(groupId, criteria)!;
+
+	return {...group.items[index]};
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/getItem.ts:16]: :thinking: why do we need to clone the item?

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+
+export default function KeyboardMovementManager({contributors, onMove}: Props) {
+	const source = useMovementSource()!;
+	const target = useMovementTarget();
+
+	const keymapRef = useRef<Record<string, KeyData>>({});
+
+	const disableMovement = useDisableKeyboardMovement();
+	const setTarget = useSetMovementTarget();
+	const setText = useSetMovementText();
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:61]: I would move all these declarations up, together with `source` and `target`.

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+	useEventListener(
+		'keydown',
+		(event) => {
+			event.stopPropagation();
+			event.preventDefault();
+
+			const {code} = event as KeyboardEvent;
+
+			const shortcut = Object.values(keymapRef.current).find(
+				(shortcut) => shortcut.keyCode === code
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:225]: `keyCode` is deprecated, we should be using `key` instead:
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+
+	const keymapRef = useRef<Record<string, KeyData>>({});
+
+	const disableMovement = useDisableKeyboardMovement();
+	const setTarget = useSetMovementTarget();
+	const setText = useSetMovementText();
+
+	const criteria = getCriteria(contributors, source.propertyKey);
+
+	keymapRef.current = {
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:65]: I guess this is being declared this way to give a name to each different keyboard action. But this is the same that using a chain of `if...else` statements in the listener:

```
// Disable movement

if (event.key === 'Escape') {
  // mimimi
}

// Execute action

else if (key === 'Enter') {
  // mimimi...
}

...
```

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+					setText(
+						sub(
+							Liferay.Language.get(
+								'targeting-x-of-item-x-of-group-x'
+							),
+							[
+								nextTarget.position,
+								`${nextTarget.index}`,
+								getGroupNumber(nextTarget.groupId)!,
+							]
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:127]: We can pass a list of parameters as arguments to the sub function:

`sub(key, param1, param2, param3)`

That way we do not need to create an array with them :smile:

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+					setTarget(nextTarget);
+
+					setText(
+						sub(
+							Liferay.Language.get(
+								'targeting-x-of-item-x-of-group-x'
+							),
+							[
+								nextTarget.position,
+								`${nextTarget.index}`,
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:125]: Prefer converting things to string using `.toString()` directly if possible, it is simpler to execute and more explicit. String literals are useful to concatenate values or handle strange cases, but for just a simple conversion this should be more than enough.

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+	const source = useMovementSource()!;
+	const target = useMovementTarget();
+
+	const keymapRef = useRef<Record<string, KeyData>>({});
+
+	const disableMovement = useDisableKeyboardMovement();
+	const setTarget = useSetMovementTarget();
+	const setText = useSetMovementText();
+
+	const criteria = getCriteria(contributors, source.propertyKey);
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:63]: if we now that this should never be null we can just put `!` here. I've seen it through all the file :sweat_smile:

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+	);
+
+	useEffect(() => {
+		if (target) {
+			return;
+		}
+
+		const initialTarget = getInitialTarget(source, criteria)!;
+
+		if (initialTarget && isValidTarget(initialTarget, source)) {
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:243]: if initial target is not null we do not need to add it to the if statement :thinking:

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+) {
+	const keyContributors = contributors.find(
+		(contributor) => contributor.propertyKey === propertyKey
+	);
+
+	if (!keyContributors) {
+		return null;
+	}
+
+	return keyContributors.criteriaMap;
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:279]: We can compress this using the `?` operator.

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+				return parentGroup;
+			}
+		}
+	}
+}
+
+function getGroupNumber(groupId: Criteria['groupId']) {
+	const splits = groupId.split('_');
+
+	return splits.pop();
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:304]: This is a one liner: `.split().pop()`

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+}
+
+function isValidTarget(target: Target, source: Source) {
+	if (target.groupId === source.groupId && target.index === source.index) {
```
Comment (p2kmgcl) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:308]: just `return !(mimimi === momomo && mumumu)`

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/cleanCriteria.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import {Criteria, CriteriaItem} from '../../types/Criteria';
+
+export default function cleanCriteria(
```
[the remark this answers is not in the document]
Comment (sandrodw3) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/cleanCriteria.ts:8]: Adding a comment with description 🚀

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/searchAndUpdateCriteria.ts (hunk 1 above)
Comment (sandrodw3) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/searchAndUpdateCriteria.ts:17]: Also adding a comment 📜

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/getItem.ts (hunk 3 above)
Comment (sandrodw3) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/getItem.ts:16]: Good question, I don't remember 🤔 I'm removing it and testing again.

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts (hunk 5 above)
Comment (sandrodw3) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:225]: Uooooh changing it

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts (hunk 6 above)
Comment (sandrodw3) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:65]: Yeah, refactoring it with KeyboardKey type as discussed in Slack 😄

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts (hunk 7 above)
Comment (sandrodw3) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:127]: Uuuh didn't know it

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts (hunk 10 above)
Comment (sandrodw3) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/keyboard_movement/KeyboardMovementManager.ts:243]: That's right 😄

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaRowEditable.js [hunk 15]
```diff
@@ -181,11 +187,31 @@ export default function CriteriaRowEditable({
...
+				}}
+				onClick={() =>
+					setMovementSource({
+						groupId,
+						icon: itemIcon,
+						index,
+						label: item.label,
+						propertyKey,
+					})
+				}
```
[the remark this answers is not in the document]
Comment (sandrodw3) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaRowEditable.js]: Jajajaja done ✅

**Comment by alessandromalves** [comment]
The failures are not related to this PR [(see comparison with acceptance routine)](https://testray.liferay.com/home/-/testray/runs?testrayBuildId=2470333899). See https://github.com/brianchandotcom/liferay-portal/pull/144828

**Comment by alessandromalves** [comment]
ci:close

**Pull request closed without merging**